### PR TITLE
Fix and guard against various alg memory leaks

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**
This follows the C++ Core Guidelines to help identify leaking
destructors:

- Mark all algorithms as final
Most algorithms are standalone, and this completely removes the need for
developers and compilers to consider inheritance

- Mark destructor for all inherited algs as protected
This allows us to check if the virtual destructor is required, by
preventing callers with the base pointer calling the destructor
implicitly (this will result in a compile time error). See cppcore
guidelines:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-dtor-virtual

- Mark remaining destructors as virtual
By using the above technique, we can spot derived types being deleted
via their base pointer without a virtual d'tor. These will leak as we
will release base memory and not the dervied instance

**To test:**
- Check we compile and tests pass
---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/inc/MantidAlgorithms/CalculateCountRate.h [hunk 1]
```diff
@@ -41,6 +42,7 @@ class MANTID_ALGORITHMS_DLL CalculateCountRate : public API::Algorithm {
   /// helper function to test if visualization workspace is requested
   bool buildVisWS() const;
 
+protected:
```
Comment (ConorMFinn) [inline Framework/Algorithms/inc/MantidAlgorithms/CalculateCountRate.h:45]: Protected still needed here?

File: Framework/Algorithms/inc/MantidAlgorithms/TOFSANSResolution.h [hunk 2]
```diff
@@ -43,6 +43,7 @@ class MANTID_ALGORITHMS_DLL TOFSANSResolution : public API::Algorithm {
 public:
   /// Defatult constructor
   TOFSANSResolution();
+  virtual ~TOFSANSResolution() = default;
```
Comment (ConorMFinn) [inline Framework/Algorithms/inc/MantidAlgorithms/TOFSANSResolution.h:46]: Comment here, to keep with the style?

**Review by cailafinn — CHANGES_REQUESTED** [review]
Good to get this done, just spotted a couple of minor things

**Review by cailafinn — APPROVED** [review]
LGTM!

File: Framework/DataHandling/inc/MantidDataHandling/SaveCanSAS1D.h [hunk 3]
```diff
@@ -80,6 +80,7 @@ Required properties:
 */
 class DLLExport SaveCanSAS1D : public API::Algorithm {
 public:
+  virtual ~SaveCanSAS1D() = default;
```
Comment (martyngigg) [inline Framework/DataHandling/inc/MantidDataHandling/SaveCanSAS1D.h:83]: There are couple of places where the algorithms are not `final` but it doesn't look they are used as bases. Have I got that wrong here and this is used somewhere along with the others?

Comment (DavidFair) [reply Framework/DataHandling/inc/MantidDataHandling/SaveCanSAS1D.h:83]: In this case it's a weird dependency tree from the V2 alg:
https://github.com/mantidproject/mantid/blob/01916cddab80834251ab4fca9aa58389ae6c69f4/Framework/DataHandling/inc/MantidDataHandling/SaveCanSAS1D2.h#L63

I've re-applied final to all algorithms and manually removed the keyword from the ones which the compiler flags to be 100% sure though (there's only ~20-30 so it was faster to do by hand)

**Pull request merged**